### PR TITLE
DM-43404: Create diffim QA metrics

### PR DESCRIPTION
**Pull request by isullivan**

### CODE REVIEW

File: python/lsst/ip/diffim/detectAndMeasure.py [hunk 1]
```diff
@@ -547,13 +642,27 @@ def measureForcedSources(self, diaSources, science, wcs):
...
 
-    def calculateMetrics(self, difference):
+    def calculateMetrics(self, difference, diaSources, science, matchedTemplate, idFactory):
         """Add image QA metrics to the Task metadata.
 
         Parameters
         ----------
         difference : `lsst.afw.image.Exposure`
             The target image to calculate metrics for.
+        diaSources : TYPE
```
Comment (ebellm) [inline python/lsst/ip/diffim/detectAndMeasure.py:652]: Docstrings need updates in this section

File: python/lsst/ip/diffim/detectAndMeasure.py [hunk 2]
```diff
@@ -156,8 +170,27 @@ class DetectAndMeasureConfig(pipeBase.PipelineTaskConfig,
...
+        dtype=str,
+        doc="List of mask planes to include in metrics",
+        default=('BAD', 'CLIPPED', 'CR', 'DETECTED', 'DETECTED_NEGATIVE', 'EDGE',
+                 'INEXACT_PSF', 'INJECTED', 'INJECTED_TEMPLATE', 'INTRP', 'NOT_DEBLENDED',
+                 'NO_DATA', 'REJECTED', 'SAT', 'SAT_TEMPLATE', 'SENSOR_EDGE', 'STREAK', 'SUSPECT',
+                 'UNMASKEDNAN',
+                 ),
+    )
+    metricSources = pexConfig.ConfigurableField(
+        target=SkyObjectsTask,
```
Comment (ebellm) [inline python/lsst/ip/diffim/detectAndMeasure.py:186]: I wonder if using sky objects as our probe will cause us to underestimate the mask fractions?  But as we discussed it's a reasonable place to start.

File: python/lsst/ip/diffim/detectAndMeasure.py [hunk 3]
```diff
@@ -567,6 +676,65 @@ def calculateMetrics(self, difference):
...
+        src.set('y', peak['i_y'])
+        src.setCoord(science.wcs.pixelToSky(peak['i_x'], peak['i_y']))
+        selectSources = diaSources[bbox.contains(diaSources.getX(), diaSources.getY())]
+        if self.config.doSkySources:
+            selectSources = selectSources[~selectSources["sky_source"]]
+        sourceDensity = len(selectSources)/area
+        dipoleSources = selectSources[selectSources["ip_diffim_DipoleFit_flag_classification"]]
+        dipoleDensity = len(dipoleSources)/area
+        if dipoleSources:
+            meanDipoleOrientation = _angleMean(dipoleSources["ip_diffim_DipoleFit_orientation"])
```
Comment (ebellm) [inline python/lsst/ip/diffim/detectAndMeasure.py:722]: The docstring above said this was relative to the parallactic angle--I don't see that computation here, am I missing something?

File: python/lsst/ip/diffim/detectAndMeasure.py [hunk 4]
```diff
@@ -653,3 +821,13 @@ def run(self, science, matchedTemplate, difference, scoreExposure,
 
         return self.processResults(science, matchedTemplate, difference, sources, idFactory,
                                    positiveFootprints=positives, negativeFootprints=negatives)
+
+
+def _angleMean(angles):
```
Comment (ebellm) [inline python/lsst/ip/diffim/detectAndMeasure.py:826]: a test of this function would help confirm we got the math right....

**Review by ebellm — CHANGES_REQUESTED** [review]
Some clarifications here and there

File: python/lsst/ip/diffim/detectAndMeasure.py [hunk 5]
```diff
@@ -247,6 +281,62 @@ def __init__(self, **kwargs):
         for flag in self.config.badSourceFlags:
             if flag not in self.schema:
                 raise pipeBase.InvalidQuantumError("Field %s not in schema" % flag)
+
+        if self.config.doWriteMetrics:
```
Comment (ebellm) [inline python/lsst/ip/diffim/detectAndMeasure.py:285]: As I said on the analysis_tools PR, I'm worried about calling these "metrics" as well when they are stored and handled quite differently from both `lsst.verify.Metrics` and `analysis_tools` metrics.

File: python/lsst/ip/diffim/detectAndMeasure.py (hunk 3 above)
Comment (isullivan) [reply python/lsst/ip/diffim/detectAndMeasure.py]: I originally thought it should be relative to parallactic angle, but then decided to simplify the calculation. I'll update the docstring.

File: python/lsst/ip/diffim/detectAndMeasure.py (hunk 5 above)
Comment (isullivan) [reply python/lsst/ip/diffim/detectAndMeasure.py:285]: Changed to `spatiallySampledMetrics`

**Review by ebellm — APPROVED** [review]

**Pull request merged**